### PR TITLE
Converted Ingestor to FB function and wrote script to invoke it with …

### LINKED ISSUE #59: Convert Ingestor to Firebase Function
As a developer/researcher working with IoT data,
I want to POST files with sensor data to the Ingestor
So that the Agile team's application can take it in, display it store the data
Part of #43

**Acceptance Criteria**

 - New Firebase function that accepts a POST request with the CSV as the body
 - Firebase function validates the file type etc.
 - then runs the Ingestor code to parse the input data the same way as before

 - New File processing script iterates through files like before and sends them to the Ingestor via HTTP POST

**Is your feature request related to a problem? Please describe.**
Currently, the Ingestor is only run locally - in a sense, it isn't really an Ingestor so much as data generator / uploader. It isn't in the cloud so it can't accept IoT data

**Describe the solution you'd like**
The Ingestor should be an HTTP Firebase Function with Express. It should accept a POST request with a file as the body. It should use this file exactly like it does currently (reads in values, scans, calculates RMS, does threshold detection, etc.)

**Describe alternatives you've considered**
Using Google Cloud Platform IoT core would be great for hooking-up data, but to learn and integrate it into our project would be beyond the time framem

**Additional context**
Partially implements #43

### PR DESCRIPTION
**Pull request by margeobur**
## GitHub Issue Solved:

closes #59 

## Current behaviour

Currently, the Ingestor is just a script that gets run locally. It reads files from the local disk, parses them and runs them through all the Ingestor logic (RMS calculation, firestore upload, etc). The left side of the figure in the Notes section shows what this looks like.

This means whoever wants to interface with our application has to have this whole script or something similar running to get data into our database and send notifications. This requires having the Node Firebase SDK etc. installed. While easily managed through NPM, this is an undesirable dependency to enforce at the data source.

## Changed behaviour

The code that does the CSV parsing, RMS calculation, threshold detection, Firestore storage and notification sending is all part of a Firebase Cloud Function now. This Function exposes a single HTTP POST endpoint that accepts a file upload of content type 'text/csv'. The path of the URL for this endpoint contains the machine and sensor IDs, and the timestamp is sent as one of the HTTP headers. The right-hand side of the diagram in the notes section shows what this looks like.

The Ingestor uses much of the same code as before. The file format that it accepts is exactly the same as the example files we've been given. This, to me, seemed like the safest option for what kind of data format the Ingestor should accept.

## PR checklist

Remember to check the following

 - [x] Tag specific people to review your PR
 - [x] Update the ReadMe with any new run instructions
 - [x] Closes the associated issue (if applicable)
 - [x] Updated kanban board

## Notes
Ingestor_Before_After-2

### Caveats

**THIS IS IMPORTANT**

There is a caveat in putting the code inside a Firebase Function. The global state isn't guaranteed to be persisted between invocations. This object: 

Screen Shot 2020-10-01 at 12 01 54 PM

contains the state that has to be "remembered" between each input data file for a single sensor. However, with a Firebase Function, the value of that object will possibly be lost between processing each file. The most important effect of this is that array of RMS values is lost. This means you have to read them in from Firestore. I know that sounds like a pain but it's worth it so that the Ingestor is running in the cloud and exposes an HTTP interface. It is also desireable, in a sense, because it means we will only be consuming Cloud Function memory while the Ingestor is servicing requests from wherever the data is coming from. Remember, these files will only come in once every 10 minutes, so we don't really need to store the RMS values in memory because that's quite inefficient - reading them from Firestore every time isn't that expensive.

Running the Ingestor locally, the state *does* get retained, so it works. @Deagler would have to verify my suspicion that we need to code the Ingestor in a stateless way if we want it to work when it's deployed.

The HTTP endpoint currently accepts the machineID and sensorID as parameters, as mentioned. But it hasn't been coded to actually support multiple machines or sensors. This will be described in an issue (or @AlexanderTheGrape may want to just talk to me to find out about that, since the day is running on).

### Demo

terminals

### Run instructions

To run, do what you'd normally do with `npm i`. Before you run, make a copy of `api/functions/env.sample` and call it `.env`. For this you can get rid of `GOOGLE_APPLICATION_CREDENTIALS` (from that file) - we don't need it any more.

Run `npm start` from either the root of the project or `api/functions` - your choice. Once that's going, open up a new cmd prompt / terminal and run submitFiles.ts with ts-node e.g. `ts-node src/ingestor/submitFiles.ts` if I'm inside `api/functions`.

Unless I've forgotten something, this should just start populating sensor 'KJPGkuNVivC5scheOIz0' of machine '2hKLutzjE0ufQen8xTm3' with data. You should see both the submitFiles.ts script and the Ingestor FB Function log stuff in the console.

### CODE REVIEW

**Comment by Deagler** [comment]
damn bro i been playing checkers while u out here playing chess

**Comment by Deagler** [comment]
Also you're right. The state should be in Firestore - it won't take a lot of work. You can make a separate collection with a single document containing that data and just retrieve/update it when needed. 

Don't worry about concurrency/transactions.

File: api/functions/src/downloads/server.ts [hunk 1]
```diff
@@ -8,8 +8,12 @@ export function ConstructFileDownloadServer() {
...
 
   app.get('/machine/:machineId/sensor/:sensorId', async (req, res) => {
-    if (!req.params.sensorId || !req.params.machineId) {
-      res.status(400).send('Bad Request');
+    if (!req.params.sensorId) {
+      res.status(400).send('No sensorId provided');
+      return;
+    }
+    if (!req.params.machineId) {
+      res.status(400).send('No machineId provided');
```
Comment (margeobur) [inline api/functions/src/downloads/server.ts:16]: Technically this addition shouldn't be in this PR but it doesn't matter - I'm just separating out the checks and sending more helpful error messages

Comment (AlexanderTheGrape) [reply api/functions/src/downloads/server.ts:16]: Good spotting, that existing check wasn't helpful for debugging.

**Review by AlexanderTheGrape — APPROVED** [review]
Tested it following the instructions, worked as expected. Very well done! The PR has been very well written too. I updated the readme with instructions and moved the story to "done" kanban board to tick off the last checkboxes.

**Pull request merged**